File: touch/src/main/java/de/nicidienase/chaosflix/touch/fragments/EventDetailsFragment.java
package de.nicidienase.chaosflix.touch.fragments;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.v7.widget.Toolbar;
import android.transition.Transition;
import android.transition.TransitionInflater;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.squareup.picasso.Callback;
import com.squareup.picasso.Picasso;

import de.nicidienase.chaosflix.R;
import de.nicidienase.chaosflix.common.entities.recording.persistence.PersistentEvent;
import de.nicidienase.chaosflix.common.entities.recording.persistence.PersistentRecording;
import de.nicidienase.chaosflix.common.entities.userdata.WatchlistItem;
import de.nicidienase.chaosflix.databinding.FragmentEventDetailsNewBinding;
import de.nicidienase.chaosflix.touch.Util;

public class EventDetailsFragment extends BrowseFragment {
	private static final String TAG = EventDetailsFragment.class.getSimpleName();
	private static final String EVENT_PARAM = "event_param";

	private OnEventDetailsFragmentInteractionListener listener;

	private long eventId;
	private boolean appBarExpanded;
	private PersistentEvent event;
	private WatchlistItem watchlistItem;

	public static EventDetailsFragment newInstance(long eventId) {
		EventDetailsFragment fragment = new EventDetailsFragment();
		Bundle args = new Bundle();
		args.putLong(EVENT_PARAM, eventId);
		fragment.setArguments(args);
		return fragment;
	}

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setHasOptionsMenu(true);
		postponeEnterTransition();
		Transition transition = TransitionInflater.from(getContext())
				.inflateTransition(android.R.transition.move);
//		transition.setDuration(getResources().getInteger(R.integer.anim_duration));
		setSharedElementEnterTransition(transition);

		if (getArguments() != null) {
			eventId = getArguments().getLong(EVENT_PARAM);
		}
	}

	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container,
							 Bundle savedInstanceState) {
		return inflater.inflate(R.layout.fragment_event_details_new, container, false);
	}

	@Override
	public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
		super.onViewCreated(view, savedInstanceState);
		FragmentEventDetailsNewBinding binding = FragmentEventDetailsNewBinding.bind(view);
		binding.playFab.setOnClickListener(v -> {
			play();
		});
		if (listener != null)
			listener.setActionbar(binding.animToolbar);

		binding.appbar.addOnOffsetChangedListener((appBarLayout, verticalOffset) -> {
			double v = (double) Math.abs(verticalOffset) / appBarLayout.getTotalScrollRange();
			if (appBarExpanded ^ v > 0.8) {
				if (listener != null) {
					listener.onToolbarStateChange();
				}
				appBarExpanded = v > 0.8;
				binding.collapsingToolbar.setTitleEnabled(appBarExpanded);
			}
		});

		getViewModel().getEventById(eventId)
				.observe(this, event -> {
					this.event = event;
					updateBookmark();
					binding.setEvent(event);
					binding.thumbImage.setTransitionName(getString(R.string.thumbnail) + event.getEventId());
					Picasso.with(getContext())
							.load(event.getThumbUrl())
							.noFade()
							.into(binding.thumbImage, new Callback() {
								@Override
								public void onSuccess() {
									startPostponedEnterTransition();
								}

								@Override
								public void onError() {
									startPostponedEnterTransition();
								}
							});
				});
	}

	private void updateBookmark() {
		getViewModel().getBookmarkForEvent(eventId)
				.observe(this,watchlistItem -> {
					this.watchlistItem = watchlistItem;
					listener.invalidateOptionsMenu();
				});
	}

	private void play() {
		if (listener != null && event != null) {
			getViewModel().getRecordingForEvent(eventId)
					.observe(this, persistentRecordings
							-> listener.playItem(event,
							Util.INSTANCE.getOptimalStream(persistentRecordings)));
		}
	}

	@Override
	public void onAttach(Context context) {
		super.onAttach(context);
		if (context instanceof OnEventDetailsFragmentInteractionListener) {
			listener = (OnEventDetailsFragmentInteractionListener) context;
		} else {
			throw new RuntimeException(context.toString()
					+ " must implement OnFragmentInteractionListener");
		}
	}

	@Override
	public void onDetach() {
		super.onDetach();
		listener = null;
	}

	@Override
	public void onPrepareOptionsMenu(Menu menu) {
		Log.d(TAG,"OnPrepareOptionsMenu");
		super.onPrepareOptionsMenu(menu);
		if(watchlistItem != null){
			menu.findItem(R.id.action_bookmark).setVisible(false);
			menu.findItem(R.id.action_unbookmark).setVisible(true);
		} else {
			menu.findItem(R.id.action_bookmark).setVisible(true);
			menu.findItem(R.id.action_unbookmark).setVisible(false);
		}
	}

	@Override
	public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
		super.onCreateOptionsMenu(menu, inflater);
//		if (appBarExpanded)
		inflater.inflate(R.menu.details_menu, menu);

	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		switch (item.getItemId()) {
			case R.id.action_play:
				play();
				return true;
			case R.id.action_bookmark:
				getViewModel().createBookmark(eventId);
				updateBookmark();
				return true;
			case R.id.action_unbookmark:
				getViewModel().removeBookmark(eventId);
				watchlistItem = null;
				listener.invalidateOptionsMenu();
				return true;
			case R.id.action_download:
				Snackbar.make(getView(), "Download not yet implemented", Snackbar.LENGTH_LONG).show();
				return true;
			default:
				return super.onOptionsItemSelected(item);
		}
	}

	public interface OnEventDetailsFragmentInteractionListener {
		void onToolbarStateChange();
		void setActionbar(Toolbar toolbar);
		void invalidateOptionsMenu();
		void playItem(PersistentEvent event, PersistentRecording recording);
	}
}

File: touch/src/main/java/de/nicidienase/chaosflix/touch/eventdetails/EventDetailsFragment.kt
package de.nicidienase.chaosflix.touch.eventdetails

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestOptions
import com.google.android.material.appbar.AppBarLayout
import com.google.android.material.snackbar.Snackbar
import de.nicidienase.chaosflix.common.mediadata.entities.recording.persistence.Event
import de.nicidienase.chaosflix.common.userdata.entities.watchlist.WatchlistItem
import de.nicidienase.chaosflix.common.viewmodel.DetailsViewModel
import de.nicidienase.chaosflix.common.viewmodel.ViewModelFactory
import de.nicidienase.chaosflix.touch.OnEventSelectedListener
import de.nicidienase.chaosflix.touch.R
import de.nicidienase.chaosflix.touch.browse.adapters.EventRecyclerViewAdapter
import de.nicidienase.chaosflix.touch.databinding.FragmentEventDetailsBinding

class EventDetailsFragment : androidx.fragment.app.Fragment() {

    private var listener: OnEventDetailsFragmentInteractionListener? = null

    private var appBarExpanded: Boolean = false
    private lateinit var event: Event
    private var watchlistItem: WatchlistItem? = null
    private var eventSelectedListener: OnEventSelectedListener? = null

    private var layout: View? = null

    private lateinit var viewModel: DetailsViewModel

    private lateinit var relatedEventsAdapter: EventRecyclerViewAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
//        postponeEnterTransition()
//        val transition = TransitionInflater.from(context)
//                .inflateTransition(android.R.transition.move)
//        //		transition.setDuration(getResources().getInteger(R.integer.anim_duration));
//        sharedElementEnterTransition = transition

        if (arguments != null) {
            val parcelable = arguments?.getParcelable<Event>(EVENT_PARAM)
            if (parcelable != null) {
                event = parcelable
            } else {
                throw IllegalStateException("Event Missing")
            }
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val binding = FragmentEventDetailsBinding.inflate(inflater, container, false)
        binding.event = event
        binding.playFab.setOnClickListener { play() }
        if (listener != null) {
            (activity as AppCompatActivity).setSupportActionBar(binding.animToolbar)
            (activity as AppCompatActivity).supportActionBar?.setDisplayHomeAsUpEnabled(true)
        }

        binding.relatedItemsList.apply {
            relatedEventsAdapter = EventRecyclerViewAdapter {
                viewModel.relatedEventSelected(it)
            }
            relatedEventsAdapter.showConferenceName = true
            adapter = relatedEventsAdapter
            val columns: Int = resources.getInteger(R.integer.num_columns)
            layoutManager = if (columns == 1) {
                LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
            } else {
                StaggeredGridLayoutManager(columns, StaggeredGridLayoutManager.VERTICAL)
            }
        }

        binding.appbar.addOnOffsetChangedListener(AppBarLayout.OnOffsetChangedListener { appBarLayout, verticalOffset ->
            val v = Math.abs(verticalOffset).toDouble() / appBarLayout.totalScrollRange
            if (appBarExpanded xor (v > 0.8)) {
                listener?.onToolbarStateChange()
                appBarExpanded = v > 0.8
//                binding.collapsingToolbar.isTitleEnabled = appBarExpanded
            }
        })

        viewModel = ViewModelProviders.of(
                requireActivity(),
                ViewModelFactory.getInstance(requireContext()))
                .get(DetailsViewModel::class.java)

        viewModel.setEvent(event)
                .observe(viewLifecycleOwner, Observer {
                    Log.d(TAG, "Loading Event ${event.title}, ${event.guid}")
                    updateBookmark(event.guid)
                    binding.thumbImage.transitionName = getString(R.string.thumbnail) + event.guid

                    Glide.with(binding.thumbImage)
                            .load(event.thumbUrl)
                            .apply(RequestOptions().fitCenter())
                            .into(binding.thumbImage)
                })
        viewModel.getRelatedEvents(event).observe(viewLifecycleOwner, Observer {
            if (it != null) {
                relatedEventsAdapter.items = it
            }
            if (it?.isNotEmpty() == true) {
                binding.relatedItemsText.visibility = View.VISIBLE
            } else {
                binding.relatedItemsText.visibility = View.GONE
            }
        })

        return binding.root
    }

    private fun updateBookmark(guid: String) {
        viewModel.getBookmarkForEvent(guid)
                .observe(viewLifecycleOwner, Observer { watchlistItem: WatchlistItem? ->
                    this.watchlistItem = watchlistItem
                    listener?.invalidateOptionsMenu()
                })
    }

    private fun play() {
        viewModel.play(event)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is OnEventSelectedListener) {
            eventSelectedListener = context
        }
        if (context is OnEventDetailsFragmentInteractionListener) {
            listener = context
        } else {
            throw RuntimeException(context.toString() + " must implement OnFragmentInteractionListener")
        }
    }

    override fun onDetach() {
        super.onDetach()
        listener = null
    }

    override fun onPrepareOptionsMenu(menu: Menu) {
        Log.d(TAG, "OnPrepareOptionsMenu")
        super.onPrepareOptionsMenu(menu)
        menu.findItem(R.id.action_bookmark).isVisible = watchlistItem == null
        menu.findItem(R.id.action_unbookmark).isVisible = watchlistItem != null
// 		menu.findItem(R.id.action_download).isVisible = viewModel.writeExternalStorageAllowed
// 		viewModel.offlineItemExists(event).observe(viewLifecycleOwner, Observer { itemExists->
// 					itemExists?.let {exists ->
// 						menu.findItem(R.id.action_download).isVisible =
// 								viewModel.writeExternalStorageAllowed && !exists
// 						menu.findItem(R.id.action_delete_offline_item).isVisible =
// 								viewModel.writeExternalStorageAllowed && exists
// 					}
// 				})

        menu.findItem(R.id.action_play).isVisible = appBarExpanded
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        // 		if (appBarExpanded)
        inflater.inflate(R.menu.details_menu, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                activity?.finish()
                return true
            }
            R.id.action_play -> {
                play()
                return true
            }
            R.id.action_bookmark -> {
                viewModel.createBookmark(event.guid)
                updateBookmark(event.guid)
                return true
            }
            R.id.action_unbookmark -> {
                viewModel.removeBookmark(event.guid)
                watchlistItem = null
                listener?.invalidateOptionsMenu()
                return true
            }
            R.id.action_download -> {
                viewModel.downloadRecordingForEvent(event)
                return true
            }
            R.id.action_delete_offline_item -> {
                viewModel.deleteOfflineItem(event).observe(viewLifecycleOwner, Observer { success ->
                    if (success != null) {
                        view?.let { Snackbar.make(it, "Deleted Download", Snackbar.LENGTH_SHORT).show() }
                    }
                })
                return true
            }
            R.id.action_share -> {
                val shareIntent = Intent(Intent.ACTION_SEND, Uri.parse(event.frontendLink))
                shareIntent.putExtra(Intent.EXTRA_SUBJECT, getString(R.string.share_description))
                shareIntent.putExtra(Intent.EXTRA_TEXT, event.frontendLink)
                shareIntent.type = "text/plain"
                startActivity(shareIntent)
                return true
            }
            R.id.action_external_player -> {
                viewModel.playInExternalPlayer(event)
                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

    interface OnEventDetailsFragmentInteractionListener {
        fun onToolbarStateChange()
        fun invalidateOptionsMenu()
    }

    companion object {
        private val TAG = EventDetailsFragment::class.java.simpleName
        private const val EVENT_PARAM = "event_param"

        fun newInstance(event: Event): EventDetailsFragment {
            val fragment = EventDetailsFragment()
            val args = Bundle()
            args.putParcelable(EVENT_PARAM, event)
            fragment.arguments = args
            return fragment
        }
    }
}
